### ISSUE TITLE
How to define custom parsers for types that have no sensible default value?

### ISSUE THREAD

**Issue by NicolasRouquette** [vigoo/clipp#172]
The doc for [custom parsers](https://vigoo.github.io/clipp/docs/parsers) requires implementing a type class:

```
trait ParameterParser[T] {
  def parse(value: String): Either[String, T]
  def default: T
}
```

For example, using the [kebs tag library](https://github.com/theiterators/kebs), I can define a tagged type for an existing executable file like this:

```
import io.github.vigoo.clipp.ParameterParser
import pl.iterators.kebs.tagged._

  sealed trait ExecutableTag

  implicit val executableFileParser: ParameterParser[Option[File] @@ ExecutableTag] =
    new ParameterParser[Option[File] @@ ExecutableTag] {
      override def parse(value: String): Either[String, Option[File] @@ ExecutableTag] = {
        val f = new File(value)
        val v0 = Right(Some(f).taggedWith[ExecutableTag])
        val v1 = if (f.canExecute) v0 else Left(s"The path must be an executable file, got: $value")
        val v2 = if (f.isFile) v1 else Left(s"The path must be a file, got: $value")
        val v3 = if (f.isAbsolute) v2 else Left(s"The path must be absolute, got: $value")
        v3
      }

      override def default: Option[File] @@ ExecutableTag = None.taggedWith[ExecutableTag]
    }
```

Since it does not make sense to define a default file value, the custom parameter type is: `Option[File] @@ ExecutableTag`.

However,  when defining a custom parameters data structure, I would like to ensure we have an existing file; however, it is unclear how to turn the lack of a file into a parser error; e.g.:

```
case class Parameters(tool: File @@ ExecutableTag)

object Parameters {
   val spec: Free[Parameter, Parameters] = for {
     _ <- metadata(....)
     otool <- namedParameter[Option[File] @@ ExecutableTag](
        description = "an executable file",
        placeholder = "tool",
        shortName = 't',
        longNames = "tool")
     tool <- ??? // we would like to get otool if Some and produce an error if otool if None.
   } yield Parameters(tool)
}
```

How could we solve this problem?

**Comment by vigoo** [comment]
This is a good question. First of all, maybe I should rename `default` to `example` because what it is really used for is that when generating the usage info graph, the parser is actually executed and those are the values that the "simulation" provides to the continuations. So it is essential to have them for how clipp works. 

So I think in your use case, you should define it for `File @@ ExecutableTag` not for an `Option` one and just provide a fake file in as a `default` (if we'll rename it to `example` then it may not be that painful :) Clipp will never use that value for anything else than for the usage graph building. 

That said it would also be nice to be able to express custom failures like you described and for that we need a `fail` command for the parser DSL.

**Comment by NicolasRouquette** [comment]
Thanks for the quick reply.

Renaming `default` to `example` would be useful.

Another way to deal with this problem is to shift the analysis of the parameters to a different Z layer.

For example:

```
case class Parameters1(tool: Option[File] @@ ExecutableTag)
case class Parameters2(tool: Option[File] @@ ExecutableTag)
```

The idea would be to get `Parameters1` from Clipp and define a Z layer to convert `Parameters1` to `Parameters2` where `None` could be rejected as an error.

While this is doable, it would be nice to avoid having to create multiple parameters case classes.

**Comment by vigoo** [comment]
Released the above linked changes as version `0.5.0`. I believe it solves this particular issue so I'm closing this, please feel free to reopen/open new issues if necessary!

**Comment by NicolasRouquette** [comment]
Thanks, the combination of clipp + prox is awesome!
